benchmarks/distributed/rpc/rl/agent.py

```diff
@@ -0,0 +1,165 @@
+        self.id = rpc.get_worker_info().id
+        self.running_reward = 0
+        self.eps = 1e-7
+
+        self.ob_rrefs = []   # Observer RRef
+        self.rewards = {}
+
+        self.future_actions = torch.futures.Future()
+        self.lock = threading.Lock()
+
+        self.agent_latency_start = None
+        self.agent_latency_end = None
+        self.agent_latency = []
+        self.agent_throughput = []
+
+    def reset_metrics(self):
+        self.agent_latency_start = None
+        self.agent_latency_end = None
+        self.agent_latency = []
+        self.agent_throughput = []
+
+    def set_world(self, batch_size, state_size, nlayers, out_features, batch=True):
+        from observer import ObserverBase
+
+        self.batch = batch
+        self.policy = Policy(state_size, nlayers, out_features, self.batch)
+        self.optimizer = optim.Adam(self.policy.parameters(), lr=1e-2)
+
+        self.batch_size = batch_size
+        for rank in range(batch_size):
+            ob_info = rpc.get_worker_info(OBSERVER_NAME.format(rank + 2))
+            self.ob_rrefs.append(
+                remote(ob_info, ObserverBase))
+            self.rewards[ob_info.id] = []
+
+        self.saved_log_probs = [] if self.batch else {
+            k: [] for k in range(self.batch_size)}
+
+        self.pending_states = self.batch_size
+        self.state_size = state_size
+        self.states = torch.zeros(self.batch_size, *state_size)
+
+    @staticmethod
+    @rpc.functions.async_execution
+    def select_action_batch(agent_rref, observer_id, state):
+        self = agent_rref.local_value()
+        observer_id -= 2
+
+        if self.pending_states == self.batch_size:
+            self.agent_latency_start = time.time()
+
+        self.states[observer_id].copy_(state)
+        future_action = self.future_actions.then(
+            lambda future_actions: future_actions.wait()[observer_id].item()
+        )
+
+        with self.lock:
+            self.pending_states -= 1
+            if self.pending_states == 0:
+                self.pending_states = self.batch_size
+                probs = self.policy(self.states)
+                m = Categorical(probs)
+                actions = m.sample()
+                self.saved_log_probs.append(m.log_prob(actions).t())
+                future_actions = self.future_actions
+                self.future_actions = torch.futures.Future()
+                future_actions.set_result(actions)
+
+                self.agent_latency_end = time.time()
+
+                batch_latency = self.agent_latency_end - self.agent_latency_start
+                self.agent_latency.append(batch_latency)
+                self.agent_throughput.append(self.batch_size / batch_latency)
+
+        return future_action
+
+    @staticmethod
+    def select_action_non_batch(agent_rref, observer_id, state):
+        self = agent_rref.local_value()
+        observer_id -= 2
+
+        self.agent_latency_start = time.time()
```

Comment:
this does not need to be an instance variable (i.e., `agent_latency_start` instead of `self.agent_latency_start`), right? In that way, we won't have contention across different threads.